### DISCUSSION TITLE
Click outside behaviour?

### DISCUSSION THREAD

**Discussion by ricardo-fnd** [ReactTooltip/react-tooltip#1076] [Q&A] [answered]
Hi devs! I went on the docs, issues and PRs, and can't find a reference to this, therefore I'm creating this discussion.
Is there a way to close the tooltip, when clicking outside of it?

Thanks.

**Comment by ricardo-fnd** [comment]
I actually found this [using `click`](https://react-tooltip.com/docs/examples/events#using-click) where it says **Clicking anywhere outside the anchor element will close the tooltip.** but this is deprecated.

**Comment by gabrieljablonski** [comment]
Hi @ricardo-fnd.

Whenever you run into deprecation warnings, you often will find more info on how to proceed close by.

![SmartSelect_20230802_073017_Edge](https://github.com/ReactTooltip/react-tooltip/assets/21102974/b1f99169-2ff9-4d1f-a96a-4c2f1b3e748e)

Is `openOnClick` what you're looking for?

**Reply by ricardo-fnd** [reply]
I'm using `openOnClick` but that prop doesn't close the tooltip when clicking outside

**Reply by gabrieljablonski** [reply]
It does in fact close the tooltip when clicking *outside the anchor*.

https://stackblitz.com/edit/stackblitz-starters-8lx6xg?file=src%2FApp.tsx

Maybe you're expecting the tooltip to close when clicking the anchor a second time, but that's not the default behavior. If that's the case you'll have to control `isOpen` manually. [Check the docs](https://react-tooltip.com/docs/examples/state) for more info on how to achieve that.

If that's not what's happening, you'll have to provide a more detailed example.

**Reply by ricardo-fnd** [reply]
I found out that when `isOpen` is `true` clicking outside doesn't work, which makes sense.
I have a problem here. I have **30 divs** which I am rendering inside a `.map()`.

When clicking in one of those divs, its children changes, and I want the tooltip to render right away, that why `isOpen` is `true`

```javascript
<div id={`tooltip-${arr.id}`} onClick={() => setActive(arr.id)}>
    {isActive ? (
        <>
            <Tooltip openOnClick clickable isOpen anchorSelect={`tooltip-${arr.id}`}>
                <p>this is a tooltip</p>
             </Tooltip>
             <div>this is the active div</div>
        </>
    ) : (
          <div>this is a div</div>
    )}
</div>
```

Any suggestions to close the tooltip when clicking outside? Thanks @gabrieljablonski

**Reply by gabrieljablonski** [reply]
Have a look at [this issue](https://github.com/ReactTooltip/react-tooltip/issues/1053#issuecomment-1618526371).

You may be able to control the tooltip target this way. Remember to use the `key` prop to make sure the tooltip gets re-rendered.

Even if in your use case you're not actually rendering the 30 tooltips at the same time, rendering them conditionally is probably not a good idea.

Try the example from the issue, placing the tooltip where it will always be on the DOM.

**Reply by gabrieljablonski** [reply, marked as answer]
As for closing the tooltip when clicking outside, you'll have to control the `isOpen` prop manually, and add an event listener to your divs to know when you've clicked outside them.

You should be able to find a generic solution to this by searching for "click outside event javascript" or something similar.

**Reply by ricardo-fnd** [reply]
So now its solved. I was able to render the tooltip only once (outside the `.map()`) and without any condition, making the `isOpen` prop the source of truth.

As I am controlling the tooltip manually by passing `isOpen` as a prop, I had to close the tooltip also manually. (Luckily I had already that implementation `useOnClickOutside`)

You might want to add that in the documentation. Passing `isOpen` as a prop, will disable click outside behaviour, which again, makes sense but it's not the first thing that comes to mind.

thanks once again @gabrieljablonski
